stock_picking_batch_creation/wizards/make_picking_batch.py

```diff
@@ -0,0 +1,305 @@
+            ("batch_id", "=", False),
+            ("user_id", "=", user.id if user else False),
+        ]
+        if not user:
+            domain.append(("printed", "=", False))
+        return domain
+
+    def _candidates_pickings_to_batch(self, user=None):
+        candidates_pickings = self._search_pickings(user=user)
+        if not candidates_pickings:
+            candidates_pickings = self._search_pickings()
+        candidates_pickings.sorted(lambda pick: pick.partner_id.id)
+        return candidates_pickings
+
+    def _compute_device_to_use(self, picking):
+        available_devices = self.stock_device_type_ids.sorted(lambda d: d.sequence)
+        # disable prefetch to ensure we compute weight and volume only
+        # for one picking at a time
+        picking = picking.with_prefetch()
+        picking._init_dimension_fields()
+
+        for device in available_devices:
+            if (
+                self._volume_condition_for_device_choice(
+                    device.min_volume,
+                    picking.total_volume_batch_picking,
+                    device.max_volume,
+                )
+                and tools.float_compare(
+                    device.max_weight,
+                    picking.total_weight_batch_picking,
+                    precision_digits=self._precision_volume(),
+                )
+                > 0
+            ):
+                return device
+        return None
+
+    def _volume_condition_for_device_choice(
+        self, min_volume, picking_volume, max_volume
+    ):
+        lower_bound = tools.float_compare(picking_volume, min_volume, 2) >= 0
+        upper_bound = tools.float_compare(max_volume, picking_volume, 2) >= 0
+        return lower_bound and upper_bound
+
+    def _get_first_suitable_device_and_picks(self, pickings):
+        """Get the first suitable device and the list of pickings where all the
+        pickings before the first one providing a suitable device are removed
+        """
+        selected_device = None
+        suitable_pick_ids = []
+        for pick in pickings:
+            if not self._lock_selected_picking(pick):
+                continue
+            device = self._compute_device_to_use(pick)
+            if device and not selected_device:
+                selected_device = device
+            if device and device == selected_device:
+                suitable_pick_ids.append(pick.id)
+
+        if selected_device:
+            return (
+                selected_device,
+                self.env["stock.picking"].browse(suitable_pick_ids),
+            )
+        return (
+            self.env["stock.device.type"].browse(),
+            self.env["stock.picking"].browse(),
+        )
+
+    def _create_batch(self, raise_if_not_possible=False):
+        user = self.user_id if self.user_id else self.env.user
+        candidates_pickings_to_batch = self._candidates_pickings_to_batch(user=user)
```

Comment:
@lmarion-source We should change the implementation to avoid to retrieve all the candidate picking in one query and making the filtering into odoo...
The idea is to:

1. Make one query to get the first picking suitable with one of our device
2. While the device is not full, make 1 query at a time to get a picking to add to the batch

With this kind of approach, we should limit the concurrent update since the result of a query should only return a picking that will be used into the batch.

To implement this approach we need:

1. A method used to get the base domain used to get a pciking candidate. This domain will be used to get the first picking and the followings.
2. A method used to get a domain to apply to filter a picking for a given device. Basically, a picking can be put into a specific device if the weight and the volume fit to the given device. The weight on the picking is already a computed stored field. If it's not case, the volume should also be a computed stored field. (The `_init_dimension_fields` methods must be removed and these fields should be declared as computed since the improvments into the orm in V16 not more justify this kind of hack)
3. A method to get the domain to apply to get a picking to add to the cluster. This method will recive as paramter the first picking retrieved, the actual cluster informations (current weight, number of lines, number of slots used, ....), the device used, ???) The challenge here is to be able to express all the criteria to apply to the picking candidates into a odoo domain...

The `_create_batch` method implementation will be therefore slightly different.

1. The first picking will retieved by querying the orm with the common domain for all pickings cancidates combinated using a *AND* operator with a domain for the device criteria. The domain for the device crieria will be a combination using a *OR* operator of the domain computed by the dedicated method for each device that can be used by the cluster.
2. Once we have the first picking and device, and while the cluster is not full, we can query for additional pickings by combining with an *AND* operator the common domain and the result to the call to the new method to get the domain to apply to add picking to the cluster...